src/tvp/feature.ts

```diff
@@ -0,0 +1,142 @@
+import { BaseLanguageClient, DynamicFeature, RegistrationData, Event } from 'coc.nvim'
+import {
+  ClientCapabilities, RPCMessageType, ServerCapabilities, Emitter,
+  ExecuteCommandParams, ExecuteCommandRequest, TextDocumentPositionParams, TextDocument, Position,
+  NotificationType, RequestType, Disposable
+} from 'vscode-languageserver-protocol'
+import {
+  TreeViewChildrenParams, TreeViewChildrenResult, TreeViewDidChangeParams,
+  TreeViewNodeCollapseDidChangeParams, TreeViewVisibilityDidChangeParams, TreeViewNode,
+  MetalsTreeRevealResult
+} from './domain'
+import { commands } from 'coc.nvim'
+import { TreeViewProvider } from './provider'
+
+interface TreeViewParams { }
+
+interface TreeViewOptions { }
+
+export class TreeViewFeature implements DynamicFeature<TreeViewParams> {
+
+  private requestType =
+    new RequestType<TreeViewParams, any, void, TreeViewOptions>('metals/treeView')
+  private treeViewChildrenParamsType =
+    new RequestType<TreeViewChildrenParams, TreeViewChildrenResult, void, void>('metals/treeViewChildren')
+  private treeViewRevealType =
+    new RequestType<TextDocumentPositionParams, MetalsTreeRevealResult, void, void>('metals/treeViewReveal')
+
+  private treeViewDidChangeType =
+    new NotificationType<TreeViewDidChangeParams, void>('metals/treeViewDidChange')
+
+  private treeViewVisibilityChangedType =
+    new NotificationType<TreeViewVisibilityDidChangeParams, void>('metals/treeViewVisibilityDidChange')
+  private treeViewNodeCollapseChangedType =
+    new NotificationType<TreeViewNodeCollapseDidChangeParams, void>('metals/treeViewNodeCollapseDidChange')
+
+  private providerEmitter: Emitter<TreeViewProvider> = new Emitter()
+  private viewUpdaters: Map<string, Emitter<TreeViewNode>> = new Map()
+  private mbGotoCommandDisposable: Disposable | undefined = undefined
+
+  constructor(private _client: BaseLanguageClient) { }
+
+  public get messages(): RPCMessageType {
+    return this.requestType
+  }
+
+  public fillClientCapabilities(capabilities: ClientCapabilities): void {
+    if (capabilities.experimental == null) {
+      capabilities.experimental = {}
+    }
+    capabilities.experimental.treeViewProvider = true
+  }
+
+  public initialize(
+    capabilities: ServerCapabilities,
+  ): void {
+    if (!capabilities.experimental!.treeViewProvider) return
+    const client = this._client
+
+    client.onNotification(this.treeViewDidChangeType, message => {
+      message.nodes.forEach(node => {
+        const viewId = node.viewId
+        const mbViewUpdater = this.viewUpdaters.get(viewId)
```

Comment:
Can we change the name of this to something else? It's not super obvious what the `mb` here is referring to, and I think it'll be much clearer if we changed it.